src/pytest_inmanta_lsm/lsm_project.py

```diff
@@ -0,0 +1,161 @@
+            },
+        )
+
+    def lsm_services_update_attributes(
+        self,
+        tid: uuid.UUID,
+        service_entity: str,
+        service_id: uuid.UUID,
+        current_version: int,
+        attributes: typing.Dict[inmanta_lsm.model.StrictStr, typing.Any],
+    ) -> inmanta.protocol.common.Result:
+        """
+        This is a mock for the lsm api, this method is called during allocation to update
+        the attributes of a service.
+        """
+        # Making some basic checks
+        service = self.services[str(service_id)]
+        assert str(tid) == self.environment, f"{tid} != {self.environment}"
+        assert service.service_entity == service_entity, f"{service.service_entity} != {service_entity}"
+        assert service.version == current_version, f"{service.version} != {current_version}"
+
+        # The attributes parameter only represents the attributes that should be changed.
+        # * When no candidate attributes were set, the new candidate attributes will be equal to the active
+        #   attributes with the attribute updates applied.
+        # * When candidate attributes were set, the update will be applied to the existing candidate
+        #   attributes.
+        if service.candidate_attributes is None:
+            service.candidate_attributes = copy.deepcopy(service.active_attributes)
+
+        service.candidate_attributes.update(attributes)
+        service.last_updated = datetime.datetime.now()
+
+        return inmanta.protocol.common.Result(code=200, result={})
+
+    def add_service(self, service: inmanta_lsm.model.ServiceInstance) -> None:
+        """
+        Add a service to the simulated environment, it will be from then one taken into account
+        in any compile.
+        """
+        if str(service.id) in self.services:
+            raise ValueError("There is already a service with that id in this environment")
+
+        self.services[str(service.id)] = service
+
+    def compile(
+        self,
+        model: str,
+        service_id: uuid.UUID,
+        validation: bool = True,
+    ) -> None:
+        """
+        Perform a compile for the service whose id is passed in argument.  The correct attribute
+        set will be selected based on the current state of the service.  If some allocation is
+        involved, the attributes of the service will be updated accordingly.
+
+        :param model: The model to compile (passed to project.compile)
+        :param service_id: The id of the service that should be compiled, the service must have
+            been added to the set of services prior to the compile.
+        :param validation_compile: Whether this is a validation compile or not.
+        """
+        service = self.services[str(service_id)]
+
+        with self.monkeypatch.context() as m:
+            m.setenv(inmanta_lsm.const.ENV_INSTANCE_ID, str(service_id))
```

Comment:
Why is there monkey patching both here and in the constructor? 

I'm a bit scared that the two parts of the monkey patching will break at different times, leading to confusion. 

Also project.compile will now be half monkeypatched?

Reply:
The monkeypatch in the constructor mocks the lsm server, the monkeypatch here mocks the environment variables passed to the compiler (which we can not pass otherwise as the compiler is in process here)

Reply:
ok, I see, Could you document this (here or at the other monkey patching site)

Reply:
Done